### DISCUSSION TITLE
How to use the slice syntax

### DISCUSSION THREAD

**Discussion by eagle-k** [xamarin/TorchSharp#982] [Q&A] [answered]
How can I write code equivalent to the following python code?

```python
tensor = torch.ones(4, 4)
print('First row: ',tensor[0])
print('First column: ', tensor[:, 0])
print('Last column:', tensor[..., -1])
tensor[:,1] = 0
print(tensor)
```

**Comment by NiklasGustafsson** [comment, marked as answer]
That depends on whether you're using .NET FX or .NET Core, since we don't have System.Range in the former. Assuming that you're using .NET Core:

```C#
var tensor = torch.ones(4, 4);
Console.WriteLine($"First row: {tensor[0].str()}");
Console.WriteLine($"First column: {tensor[.., 0].str()}");
Console.WriteLine($"Last column: {tensor[TensorIndex.Ellipsis, ^1].str()");
tensor[..,1] = 0;
Console.WriteLine(tensor.str());
```

The `str()` is necessary, because the default `ToString()` function does not print all the values, just the metadata. There's an overloaded `ToString()` that does the right thing, but I chose the `str()` extension method just for brevity's sake.

See the TorchSharp C# tutorials for more details:

https://github.com/dotnet/TorchSharpExamples/tree/main/tutorials/CSharp

**Reply by eagle-k** [reply]
I see that I can use `TensorIndex`.

By the way, there is a typo in the `Ellipsis` line, so I'll provide a modified version just in case for others.

```cs
Console.WriteLine($"Last column: {tensor[torch.TensorIndex.Ellipsis, 1].str()}");
```

The C# code worked fine on VSCode's Polyglot Notebook.
Thank you for teaching me how to do this.

I also tried to write this in F# as well as the C# version.

```fsharp
open TorchSharp
open type TorchSharp.torch

let tensor = torch.ones(4, 4)
printfn $"First row: {tensor[0].str()}"
printfn $"First column: {tensor[TensorIndex.Colon, TensorIndex.Single(0)].str()}"
printfn $"Last column: {tensor[TensorIndex.Ellipsis, TensorIndex.Single(-1)].str()}"
tensor[TensorIndex.Colon, TensorIndex.Single(1)] <- torch.tensor(0)
printfn $"{tensor.str()}"
```

The following F# slice syntax did not seem to work well.

```fsharp
tensor[*, 0] // syntax error
```

**Reply by NiklasGustafsson** [reply]
The '^1' on the Ellipsis line was intentional. I believe that's how C# would denote a -1 in an index. However, I noticed that there's no conversion from System.Index to TensorIndex, so that will have to be fixed. For now, use `shape[n]-1` where `n` is the dimension you use a negative index in.

I just filed a bug.

**Comment by NiklasGustafsson** [comment]
That said, I and didn't realize this, but you can just use -1 instead of ^1. So:

```C#
Console.WriteLine($"Last column: {tensor[TensorIndex.Ellipsis, -1].str()}");
```

**Comment by NiklasGustafsson** [comment]
The System.Index issue has been addressed in 0.99.6

**Comment by eagle-k** [comment]
It seems that F#'s slice syntax can be managed by implementing `GetSlice` and `SetSlice` separately at the user level. However, it would be even more convenient if the library could provide support for it.

```fsharp
type torch.Tensor with
    member t.GetSlice(startIdx: int option, endIdx: int option) =
        match startIdx, endIdx with
        | Some s, Some e -> t[torch.TensorIndex.Slice(s, e)]
        | Some s, None -> t[torch.TensorIndex.Slice(s)]
        | None, Some e -> t[torch.TensorIndex.Slice(System.Nullable(), e)]
        | None, None -> t[torch.TensorIndex.Slice()]

    member t.SetSlice(startIdx: int option, endIdx: int option, v: torch.Tensor) =
        match startIdx, endIdx with
        | Some s, Some e -> t[torch.TensorIndex.Slice(s, e)] <- v
        | Some s, None -> t[torch.TensorIndex.Slice(s)] <- v
        | None, Some e -> t[torch.TensorIndex.Slice(System.Nullable(), e)] <- v
        | None, None -> t[torch.TensorIndex.Slice()] <- v

let t = torch.arange(24).view(4, 3, 2)
t[1..3] <- torch.tensor(1)
printfn $"{t.str()}"
```
